oauth2/src/main/scala/tsec.oauth2/provider/GrantHandler.scala

```diff
@@ -0,0 +1,166 @@
+package tsec.oauth2.provider
+
+import cats.effect.IO
+import ca.mrvisser.sealerate
+import cats.data.EitherT
+
+import scala.concurrent.duration.FiniteDuration
+
+final case class GrantHandlerResult[U](
+                                  authInfo: AuthInfo[U],
+                                  tokenType: String,
+                                  accessToken: String,
+                                  expiresIn: Option[FiniteDuration],
+                                  refreshToken: Option[String],
+                                  scope: Option[String],
+                                  params: Map[String, String])
+
+sealed trait GrantHandler {
+  def name: String
+
+  def handleRequest[U](request: AuthorizationRequest, authorizationHandler: AuthorizationHandler[U]): IO[GrantHandlerResult[U]]
+
+  protected def getClientCredential[U](request: AuthorizationRequest, handler: AuthorizationHandler[U]): EitherT[IO, OAuthError, ClientCredential] =for {
+    credential <- EitherT.fromEither[IO](request.parseClientCredential)
+    _ <- EitherT(handler.validateClient(credential, request).map(isValid => Either.cond(isValid, (), InvalidClient("Invalid client or client is not authorized"): OAuthError)))
+  } yield credential
+
+  /**
+   * Returns valid access token.
+   */
+  protected def issueAccessToken[U](handler: AuthorizationHandler[U], authInfo: AuthInfo[U]): IO[GrantHandlerResult[U]] = {
+    handler.getStoredAccessToken(authInfo).flatMap {
+      case Some(token) if shouldRefreshAccessToken(token) => token.refreshToken.map {
+        handler.refreshAccessToken(authInfo, _)
+      }.getOrElse {
+        handler.createAccessToken(authInfo)
+      }
+      case Some(token) => IO.pure(token)
+      case None => handler.createAccessToken(authInfo)
+    }.map(createGrantHandlerResult(authInfo, _))
+  }
+
+  protected def shouldRefreshAccessToken(token: AccessToken) = token.isExpired
+
+  protected def createGrantHandlerResult[U](authInfo: AuthInfo[U], accessToken: AccessToken) = GrantHandlerResult(
+    authInfo,
+    "Bearer",
+    accessToken.token,
+    accessToken.expiresIn,
+    accessToken.refreshToken,
+    accessToken.scope,
+    accessToken.params)
+
+}
+
+object GrantHandler{
+
+  case object RefreshToken extends GrantHandler {
+    def handleRequest[U](request: AuthorizationRequest, handler: AuthorizationHandler[U]): IO[GrantHandlerResult[U]] = {
+      val res: EitherT[IO, Throwable, GrantHandlerResult[U]] = for{
```

Comment:
`EitherT[IO, Throwable]` is completely redundant, as `IO` is `MonadError[F, Throwable]`.

I agree with the use case for `EitherT` for custom error lifting. For this, just stay in `IO`.

Reply:
Same applies to the ones below this, btw.